### PR TITLE
[flash_ctrl] Vivado combo loop workaround

### PR DESCRIPTION
**Pull request by vogelpi**
By inserting a prim_flop, we can prevent Vivado from optimizing across a critical registers thereby avoiding the inferring of combo loops.

This is related to #10204 .

### CODE REVIEW

**Review by rswarbrick — APPROVED** [review]
This looks good. I think we're working around what appears to be a tool bug in Vivado, right? If so, could you add a comment to explain what's going on?

But please don't force-push until this has passed CI: assuming it does pass CI, I think we should force-push to add a comment and then immediately merge to unblock everything else (assuming there are no changes to the actual RTL). If we force-push before CI has finished, we'll have to wait for the whole run again :-(

**Comment by vogelpi** [comment]
Thanks Rupert, yes adding a comment makes sense. I'll do that after CI has passed.

I am not 100% sure it's really a bug in the tool. The `err_sts` signal is 4 bits wide and without the prim_flop Vivado inserts 2 flops plus some logic called `CARRY` (for whatever reason). The loop then goes through this logic. I think the prim flop also makes sense here even without the tool problem as the `err_sts` is relevant for security.

**Comment by vogelpi** [comment]
Hm, this is really annoying. Now we get a combo loop on the CW310 only.

**Comment by vogelpi** [comment]
After inspecting the failure on the CW310 it turns out that Vivado also inferred a combo loop through the beat counter. I've now also inserted a prim_flop for that. Let's see.

Btw after #10293 got merged, I've now rebased this on master and integrated back @tjaychen 's original commit.

File: hw/ip/flash_ctrl/rtl/flash_ctrl_lcmgr.sv
Comment (tjaychen) [inline hw/ip/flash_ctrl/rtl/flash_ctrl_lcmgr.sv:583]: thanks @vogelpi 
honestly i don't really understand why this is different and what the issue vivado is encountering..
clearly it's not what i had mentioned before...

Comment (tjaychen) [reply hw/ip/flash_ctrl/rtl/flash_ctrl_lcmgr.sv:583]: @vogelpi did you see combo loops even after the fix i made?

**Review by tjaychen — APPROVED** [review]

**Comment by tjaychen** [comment]
hmmm...looks like now it's being inferred elsewhere. There's something very wrong with this overall. 
I'm going to try and check this out locally also.  There's something in there starting from the sparse FSMs that's making vivado very unhappy in my original PR.

**Comment by tjaychen** [comment]
> Thanks Rupert, yes adding a comment makes sense. I'll do that after CI has passed.
> 
> I am not 100% sure it's really a bug in the tool. The `err_sts` signal is 4 bits wide and without the prim_flop Vivado inserts 2 flops plus some logic called `CARRY` (for whatever reason). The loop then goes through this logic. I think the prim flop also makes sense here even without the tool problem as the `err_sts` is relevant for security.

i still feel like all of this is a tool issue. Literally the only change made in the original PR is converting an fsm from vanilla to sparse. If there were loops before, they should still be there now.  I don't really see why this is changing.

**Comment by vogelpi** [comment]
Yes, this looks very much like a tool bug.

I think I've now managed to somewhat confine the bug. It seems not related to the switch to sparse FSMs itself. But with the sparse FSMs, `flash_ctrl_lcmgr.sv` can now also drive `phase` to `PhaseInvalid` as part of a real case statement (see [here](https://github.com/lowRISC/opentitan/pull/10178/files#r791751775)). Previously, this assignment was just part of the `default`. We had 9 states of which `StInvalid` wasn't defined. So I suspect Vivado just optimized it away and left 8 states resulting also in the `default` being optimized away as there is no need for a default if you have a power of two number of states.

I suspect that this led to further optimizations in the usage of `phase`. For example, inside `flash_mp.sv` it's compared against the properties obtained from the package file. And there we just set values `PhaseRma` and `PhaseSeed`. There are a couple of comparisons and many further assignments that depend on these. Eventually, the result of these comparisons feed back into `flash_ctrl_arb.sv` and then `flash_ctrl_lcmgr.sv` in the form of the `done_i` signal. @tjaychen I think this is the same signal you identified as part of the loop [here](https://github.com/lowRISC/opentitan/issues/10204#issue-1108736233)

The latest version at least synthesized locally. Let's see if it passes CI, too.

**Comment by tjaychen** [comment]
> Yes, this looks very much like a tool bug.
> 
> I think I've now managed to somewhat confine the bug. It seems not related to the switch to sparse FSMs itself. But with the sparse FSMs, `flash_ctrl_lcmgr.sv` can now also drive `phase` to `PhaseInvalid` as part of a real case statement (see [here](https://github.com/lowRISC/opentitan/pull/10178/files#r791751775)). Previously, this assignment was just part of the `default`. We had 9 states of which `StInvalid` wasn't defined. So I suspect Vivado just optimized it away and left 8 states resulting also in the `default` being optimized away as there is no need for a default if you have a power of two number of states.
> 
> I suspect that this led to further optimizations in the usage of `phase`. For example, inside `flash_mp.sv` it's compared against the properties obtained from the package file. And there we just set values `PhaseRma` and `PhaseSeed`. There are a couple of comparisons and many further assignments that depend on these. Eventually, the result of these comparisons feed back into `flash_ctrl_arb.sv` and then `flash_ctrl_lcmgr.sv` in the form of the `done_i` signal. @tjaychen I think this is the same signal you identified as part of the loop [here](https://github.com/lowRISC/opentitan/issues/10204#issue-1108736233)
> 
> The latest version at least synthesized locally. Let's see if it passes CI, too.


you know what's funny. I originally did not explicitly use `StInvalid` and just relied on the default state.  However I kept getting this inferred combo loop issue.  At the time, adding the `StInvalid` seemed to have made the problem go away in that specific PR.

**Comment by tjaychen** [comment]
@vogelpi i think it's okay to merge this. I'll follow-up with another PR to make the usage of `phase` a bit more consistent. But I think this change should still work the way we intend.

**Comment by tjaychen** [comment]
i'm going to merge this for now. Thanks @vogelpi 
I don't think we NEED to set `PhaseInvalid` as I've done, since `PhaseNone` is functionally the same.

The changes in this PR are also smaller than the experiment I ran.

**Comment by vogelpi** [comment]
Thanks @tjaychen for merging this. I was OoO yesterday.

**Pull request merged**
